docs/rfcs/0005-migrate-sitepermission-ac-store-geckoview-store.md

```diff
@@ -0,0 +1,32 @@
+---
+layout: page
+title: Migrate feature-sitepermissions to be compatible with the new GeckoView permission API.
+permalink: /rfc/0005-migrate-sitepermission-ac-store-geckoview-store
+---
+
+* Start date: 2020-07-26
+* RFC PR: [#7855](https://github.com/mozilla-mobile/android-components/pull/7855)
+
+## Summary
+
+Migrate our generic [SitePermissionsStorage](https://github.com/mozilla-mobile/android-components/blob/56e6f5ba55734014804b3f3b0af5c32414fa1d16/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsStorage.kt#L33) to be compatible with the new [GeckoView permissions API](https://docs.google.com/document/d/1KUq0gejnFm5erkHNkswm8JsT7nLOmWvs1KEGFz9FWxk/edit#heading=h.ls1dr18v7zrx).
+
+## Motivation
+
+For permissions, we currently are storing on disk all the site permissions on AC both for GeckoView and WebView, but that's about to change for GeckoView, [a new API is about to land](https://docs.google.com/document/d/1KUq0gejnFm5erkHNkswm8JsT7nLOmWvs1KEGFz9FWxk/edit#heading=h.ls1dr18v7zrx), where GeckoView is going to store all the permissions and we are going to have APIs for managing permissions(query, add, delete, and update).
+
+That put us in a situation where we have to migrate the existing user data to the new [GeckoView APIs](https://docs.google.com/document/d/1KUq0gejnFm5erkHNkswm8JsT7nLOmWvs1KEGFz9FWxk/edit#heading=h.ls1dr18v7zrx).
+
+## Reference-level explanation
+
+####  Doing a one-shot migration.
+We decide a point (when the app startup or other place) to do a background operation to transfer all the data to GeckoView in a one-shot, and until this operation is not finished all the site permissions operations will be put on hold until the migration is over.
+
+We transform [SitePermissionsStorage](https://github.com/mozilla-mobile/android-components/blob/a93eae2cbf81a17e80a98ec234db414d1f7e84f5/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsStorage.kt#L33) into an interface this way we can have two implementations one for GeckoView and the other for other engines. Inside of the GeckoView implementation we can have a preference that indicates if the migration is over or not, when the `SitePermissionsFeature.start()` gets called we start the migration and we guard all the operations(read/write) on GeckoViewStorage and hold its response until the
+migration is over.
+
+## Drawbacks
+We are going to do a fast migration but depending on how much data the user has, it could impact the user's experience, as users could load a page that requires permissions and this functionality will no be active until we finish the migration.
+
+## Unresolved questions.
+When the migration should start?
```

Comment:
Another possible question could be when can we remove this checks for the migration? AC doesn't have a way to verify if a consumer was using a previous version of AC. Then when we know that it's safe to remove the migration code

Reply:
We have a similar problem with the Fennec migration code. For site permissions I feel like we can define a fixed time frame (x months?) and then we can start to remove it. I am less concerned about site permissions since in the worst case we will just re-prompt the user and that is absolutely harmless.

Reply:
> We have a similar problem with the Fennec migration code. For site permissions I feel like we can define a fixed time frame (x months?) and then we can start to remove it. 

That could be an alternative 👍  we just have to document the time frame on the change log, for consumer to know that after that time frame we are going to re prompt.